core/src/main/scala/org/bykn/bosatsu/LetFree.scala

```diff
@@ -129,42 +180,69 @@ object LetFreeConversion {
   type LetFreePM = PackageMap.Typed[(Declaration, LetFreeExpressionTag)]
   type LetFreePac = Package.Typed[(Declaration, LetFreeExpressionTag)]
 
-  type PatternEnv = Map[Int, LetFreeExpression]
+  type PatternEnv[T] = Map[Int, T]
```

Comment:
I think using `LongMap[T]` will give significantly better performance since you won't have the `Int` boxing.